src/compiler/crystal/tools/unreachable.cr

```diff
@@ -111,6 +113,31 @@ module Crystal
         end
       end
     end
+
+    # https://docs.codecov.com/docs/codecov-custom-coverage-format
+    def to_codecov(io)
+      hits = Hash(String, Hash(Int32, Int32)).new { |hash, key| hash[key] = Hash(Int32, Int32).new(0) }
+
+      each do |a_def, location, count|
+        hits[location.filename][location.line_number] = count
+      end
```

Comment:
**thought:** This is probably fine. But it would be more efficient to simply sort defs by filename and line_number. That automatically groups them together.

**question:** Is the indexing by line number really necessary? I figure it would only have a purpose if there are duplicate entries for the same filename and line_number. But that should not be the case IIRC.

Reply:
The reasoning behind this is because the spec requires each each file to be its own object in the JSON, and each line + count to be key/value pairs in that object.

I'm open to better ideas, but this was the best way I could think of at the time. It allows iterating over the files to define that object in the JSON, then iterate again within that object to build out the key/value pairs. Something, as far as I know, you can't do with a 1D array that using `each` directly operates on.

Reply:
That should be possible. You just have to use manual `start_object`/`end_object` for the grouping boundaries.

Reply:
Sorry I'm not sure I follow. If you have an array of locations like `each` is iterating over, how would you know when to emit the `start_object`/`end_object` without keeping track of the filename of each location somehow?

Reply:
I was thinking about something like this:
```cr
@tallies.sort_by {|a_def, _| {a_def.location.not_nil!.filename.as(String), a_def.location.not_nil!.line_number} }

JSON.build(io) do |builder|
  builder.object do
    builder.string "coverage"
    builder.object do
      last_filename = nil
      each do |a_def, location, count|
        if location.filename != last_filename
          builder.end_object unless last_filename.nil?
          last_filename = location.filename
          builder.string location.filename
          builder.start_object
        end
        builder.field location.line_number, count
      end
      builder.end_object unless last_filename.nil?
    end
  end
end
```

Reply:
Ahh okay, so you do have to keep track of the filenames. Happy to push this up if we think performance is going to be an issue. Otherwise I think I'm leaning towards keeping it as is given it's a bit more readable?

Reply:
As I said, the current implementation is probably fine. Just noting there's a more efficient alternative.